airbyte-integrations/connectors/source-zuora/source_zuora/zuora_client.py

```diff
@@ -0,0 +1,270 @@
+        submit_job = self._make_request(method="POST", url=f"{self.url_base}/query/jobs", payload=dq_query)
+        # self.logger.debug(submit_job.json()["data"]["id"])
+        return submit_job.json()["data"]["id"]
+
+    # CHECK: the submited data_query_job status
+    def _check_dq_job_status(self, dq_job_id: str) -> Dict:
+        # Define the job error statuses
+        errors = ["failed", "canceled", "aborted"]
+        # Error msg: the cursor_field cannot be resolved
+        cursor_error = "Column 'updateddate' cannot be resolved"
+        # Error msg: cannot process object
+        obj_read_error = "failed to process object"
+
+        status = None
+        success = "completed"
+        while status != success:
+            response = self._make_request(url=f"{self.url_base}/query/jobs/{dq_job_id}")
+            dq_job_check = response.json()
+            status = dq_job_check["data"]["queryStatus"]
+            if status in errors and cursor_error in dq_job_check["data"]["errorMessage"]:
+                raise ZOQLQueryFieldCannotResolve
+            elif status in errors and obj_read_error in dq_job_check["data"]["errorMessage"]:
+                raise ZOQLQueryCannotProcessObject
+            elif status in errors:
+                raise ZOQLQueryFailed(response)
+        return dq_job_check
+
+    # GET: data_query_job result
+    def _get_data_from_dq_job(self, response: requests.Response) -> List:
+        dq_job_data = requests.get(response["data"]["dataFile"])
+        dq_job_data.raise_for_status()
+        data = [j.loads(line) for line in dq_job_data.text.splitlines()]
+        return data or []
+
+    # Warpper function for `Query > Submit > Check > Get`
+    def _get_data(
+        self, q_type: str = "select", obj: str = None, date_field: str = None, start_date: str = None, end_date: str = None
+    ) -> Iterable:
+        query = self._make_dq_query(q_type, obj, date_field, start_date, end_date)
+        # self.logger.debug(f"{query}")
+        submit = self._submit_dq_job(query)
+        check = self._check_dq_job_status(submit)
+        get = self._get_data_from_dq_job(check)
+        yield from get
+
+    # Warper for _get_data using date-slices as pages
+    def _get_data_with_date_slice(
+        self, q_type: str, obj: str, date_field: str, start_date: str, window_days: int, end_date: str = None
+    ) -> Iterable:
+        # Parsing input dates
+        s = pendulum.parse(start_date)
+        tz = self._get_tz(s)
+        # If there is no `end_date` as input - use now()
+        e = pendulum.parse(pendulum.now().to_datetime_string()) if end_date is None else pendulum.parse(end_date)
+        # check end_date bigger than start_date
+        if s >= e:
+            raise EndDateError
+        else:
+            # Get n of date-slices
+            n = self._get_n_slices(s, e, window_days)
+            # initiating slice_start/end for date slice
+            slice_start = s.to_datetime_string() + tz
```

Comment:
The reason of doing so using `adding a string` is we need to retrieve the correct information from the state, once we use incremental refresh. 
For the `start date` we will have something like : `2021-01-01T00:00:00-07:00` - just because the client's account has the timezone -7h.
For the `end date` we will have till now(), so this will be using: `2021-07-01T00:00:00+00:00` because now always use the UTC as a timezone, if we don't specify TZ for `end date`, we will sync data incorrectly. 

Also, the very important thing of getting the data from Zuora is it requires the next timestamp format:
`updateddate >= TIMESTAMP '2017-01-01 00:00:00 +00:00'`
the _get_tz method, gives us the correct timezone according to the input date or state, and then we push that formated string as input to the query parameter. It should stay as is.